### PR TITLE
Store sampleview coords in context menu state

### PR DESCRIPTION
**Pull request by axelboc**
I now store `e.offset<X|Y> / imageRatio` instead of just `e.offset<X|Y>` in the state, which saves `ContextMenu` from having to do this calculation in a couple of places.

I've made a couple of additional commits to move some logic around, inline some methods, etc., and to let the right-clicked shape be `undefined` in the state (instead of storing a fake shape object `{ type: 'NONE' }`).

### CODE REVIEW

File: ui/src/components/SampleView/ContextMenu.jsx [hunk 1]
```diff
@@ -486,7 +479,7 @@ export default class ContextMenu extends React.Component {
     const menuOptions = this.menuOptions();
     let optionList = [];
 
-    if (this.props.sampleID !== undefined) {
```
Comment (axelboc) [inline ui/src/components/SampleView/ContextMenu.jsx:489]: I've kept this condition for now, but I don't find it ideal. If there's no sample mounted and you right-click on a shape, you get a completely different context menu than if you right-click on the same shape and there's a sample mounted. Actions that don't require a sample to be mounted, like _Delete Line_, could very well be shown when right-clicking on a line regardless of the presence of a sample.

I would prefer for context menu actions that require a sample to be mounted to be conditionally rendered individually.

File: ui/src/components/SampleView/ContextMenu.jsx [hunk 2]
```diff
@@ -34,24 +33,6 @@ export default class ContextMenu extends React.Component {
       grid: [],
       none: [],
     };
-    let twoDPoints = [];
-
-    if (this.props.enable2DPoints) {
```
Comment (axelboc) [inline ui/src/components/SampleView/ContextMenu.jsx:39]: I've inlined the condition below.

File: ui/src/components/SampleView/ContextMenu.jsx [hunk 3]
```diff
@@ -253,32 +252,53 @@ export default class ContextMenu extends React.Component {
...
+              },
+            ]
+          : []),
         { text: 'divider', key: 7 },
         ...genericTasks.none,
-        genericTasks.grid.none > 0 ? { text: 'divider', key: 7 } : {},
       ],
     };
 
-    if (this.props.enableNativeMesh) {
```
Comment (axelboc) [inline ui/src/components/SampleView/ContextMenu.jsx:274]: Also inlined above.

File: ui/src/components/SampleView/ContextMenu.jsx (hunk 1 above)
Comment (marcus-oscarsson) [reply ui/src/components/SampleView/ContextMenu.jsx:489]: > I would prefer for context menu actions that require a sample to be mounted to be conditionally rendered individually or to show an error to the user when used.

Yes, the context menu on a shape without a sample mounted should give an error, just like we can't create a point/shape without a sample

Comment (axelboc) [reply ui/src/components/SampleView/ContextMenu.jsx:489]: Sorry, I've edited my comment to clarify my thought: I think that actions that don't require a sample to be mounted, like _Delete Point_ should be shown when right-clicking on a shape regardless of the presence of a sample.

That's because when you mount a sample, perform 3-click centring, save the created point, and unmount the sample, the point remains on the sample view, so you can still right-click on it. At the moment, when you're in this situation, you basically can't delete anything via the context menu (you still can by pressing <kbd>Delete</kbd>):

image

Comment (marcus-oscarsson) [reply ui/src/components/SampleView/ContextMenu.jsx:489]: Hm, yes makes sense from that "point" (no pun intended :) ) of view. But all the shapes should actually be cleared when a sample is unmounted, because they are not valid any longer. We should look at that if its not the case

Comment (marcus-oscarsson) [reply ui/src/components/SampleView/ContextMenu.jsx:489]: On the beamlines, at least ours, the shapes are removed at unmount, I guess its just a event that is not registered properly in some mockup.

**Comment by axelboc** [comment]
Replacing the dummy shape `{ type: 'NONE' }` with `undefined` in the context menu state seems to have caused some issues in the `showModal` method. Apologies, I'll investigate.

**Comment by axelboc** [comment]
I've used optional chaining in `showModal` to handle an undefined `shape`. In the other methods, the `shape` prop is always defined.

File: ui/src/components/SampleView/ContextMenu.jsx [hunk 4]
```diff
@@ -424,24 +423,20 @@ export default class ContextMenu extends React.Component {
...
-    const { imageX, imageY, imageRatio } = this.props;
-    this.createPoint(imageX / imageRatio, imageY / imageRatio, (shape) =>
-      this.showModal(name, {}, shape, extraParams),
+    const { sampleViewX, sampleViewY } = this.props;
+
+    this.props.sampleViewActions.add2DPoint(
+      sampleViewX,
+      sampleViewY,
+      'SAVED',
+      (shape) => this.showModal(name, {}, shape, extraParams),
```
Comment (walesch-yan) [inline ui/src/components/SampleView/ContextMenu.jsx:439]: Maybe you already thought about it and want to do that in a different PR, but related to #1490: Couldn't we just prevent the drawing of a point when no sample is mounted by adding a check for `sampleData` (similar to how it is done in `showModal`)

Comment (axelboc) [reply ui/src/components/SampleView/ContextMenu.jsx:439]: Indeed, the check for the mounted sample could just come before the creation of the point. But we could also hide or disable the context menu option.

Comment (walesch-yan) [reply ui/src/components/SampleView/ContextMenu.jsx:439]: You are correct, that might be a more elegant solution to the problem. Sorry, It seems that I misunderstood the other comment, the first time I read it
